patches/server/0722-Rate-options-and-timings-for-sensors-and-behaviors.patch

```diff
@@ -0,0 +1,217 @@
++                            table.put(entity.toUpperCase(Locale.ROOT), typeName.toUpperCase(Locale.ROOT), tickRate);
++                            log("      " + typeName + ": " + tickRate);
++                        }
++                    }
++                }
++            }
++        }
++
++        if (table.isEmpty()) {
++            log("    None configured");
++        }
++        return table;
++    }
++
++    public int getBehaviorTickRate(String typeName, String entityType, int def) {
++        return getIntOrDefault(behaviorTickRates, typeName, entityType, def);
++    }
++
++    public int getSensorTickRate(String typeName, String entityType, int def) {
++        return getIntOrDefault(sensorTickRates, typeName, entityType, def);
++    }
++
++    private int getIntOrDefault(Table<String, String, Integer> table, String rowKey, String columnKey, int def) {
++        Integer rate = table.get(columnKey, rowKey);
++        return rate != null && rate > -1 ? rate : def;
++    }
+ }
+ 
+diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/Behavior.java b/src/main/java/net/minecraft/world/entity/ai/behavior/Behavior.java
+index b1212e162ba938b3abe0df747a633ba9cbbe57c8..f689fad6f4b728f676fb1feaf742c0305dbf4dae 100644
+--- a/src/main/java/net/minecraft/world/entity/ai/behavior/Behavior.java
++++ b/src/main/java/net/minecraft/world/entity/ai/behavior/Behavior.java
+@@ -14,6 +14,10 @@ public abstract class Behavior<E extends LivingEntity> {
+     private long endTimestamp;
+     private final int minDuration;
+     private final int maxDuration;
++    // Paper start - configurable behavior tick rate and timings
++    private final String configKey;
++    private final co.aikar.timings.Timing timing;
++    // Paper end
+ 
+     public Behavior(Map<MemoryModuleType<?>, MemoryStatus> requiredMemoryState) {
+         this(requiredMemoryState, 60);
+@@ -27,6 +31,15 @@ public abstract class Behavior<E extends LivingEntity> {
+         this.minDuration = minRunTime;
+         this.maxDuration = maxRunTime;
+         this.entryCondition = requiredMemoryState;
++        // Paper start - configurable behavior tick rate and timings
++        String key = getClass().getName().startsWith("net.minecraft.") ? getClass().getSimpleName() : getClass().getName();
++        key = key.toLowerCase(java.util.Locale.ROOT);
++        if (key.startsWith("behavior")) {
++            key = key.substring("behavior".length());
++        }
++        this.configKey = key;
++        this.timing = co.aikar.timings.MinecraftTimings.getBehaviorTimings(configKey);
++        // Paper end
+     }
+ 
+     public Behavior.Status getStatus() {
+@@ -34,11 +47,19 @@ public abstract class Behavior<E extends LivingEntity> {
+     }
+ 
+     public final boolean tryStart(ServerLevel world, E entity, long time) {
++        // Paper start - behavior tick rate
++        int tickRate = world.paperConfig.getBehaviorTickRate(configKey, entity.getType().id, -1);
```

Comment:
Please use `this` when referring to class properties; that goes in line with the changes Paperweight does for us automatically, and of course makes it easier to read. Merger can handle too.